blackbird_python/blackbird/tests/test_auxiliary.py

```diff
@@ -0,0 +1,558 @@
+        func.COS = lambda: True
+        expression = num(n)
+        assert _func(func, expression) == np.cos(expected)
+
+    @pytest.mark.parametrize('n, expected', test_complex)
+    def test_function_sqrt(self, parser, ctx, n, expected, num):
+        """Test that a Blackbird sqrt function is properly called"""
+        func = blackbirdParser.FunctionContext(parser, ctx)
+        func.SQRT = lambda: True
+        expression = num(n)
+        assert _func(func, expression) == np.sqrt(expected)
+
+    def test_function_invalid(self, parser, ctx):
+        """Test that an invalid Blackbird function raises the correct exception"""
+        func = blackbirdParser.FunctionContext(parser, ctx)
+        func.getText = lambda: "Hello"
+
+        with pytest.raises(NameError, match="Unknown function Hello"):
+            _func(func, None)
+
+
+class TestExpression:
+    """Tests for the _expression function"""
+
+    @pytest.mark.parametrize('n, expected', test_complex)
+    def test_number(self, n, expected, num):
+        """Test that a Blackbird expression containing numbers evaluates"""
+        expr = num(n)
+        assert _expression(expr) == expected
+
+    def test_variable(self, parser, ctx, monkeypatch):
+        """Test that a Blackbird expression containing variables evaluates"""
+        expr = blackbirdParser.VariableLabelContext(parser, ctx)
+        expr.getText = lambda: "var1"
+
+        with monkeypatch.context() as m:
+            m.setattr(blackbird.auxiliary, "_VAR", {"var1": 5})
+            assert _expression(expr) == 5
+
+    def test_variable_invalid(self, parser, ctx, monkeypatch):
+        """Test that an error is raised if the variable does not exist"""
+        expr = blackbirdParser.VariableLabelContext(parser, ctx)
+        expr.getText = lambda: "var2"
+
+        with monkeypatch.context() as m:
+            m.setattr(blackbird.auxiliary, "_VAR", {"var1": 5})
+            with pytest.raises(NameError, match="name 'var2' is not defined"):
+                _expression(expr)
+
+    @pytest.mark.parametrize('n1', test_complex)
+    @pytest.mark.parametrize('n2', test_floats)
+    def test_brackets(self, parser, num, n1, n2, ctx):
+        """Test that brackets are correctly parsed"""
+
+        class DummyAddLabel(blackbirdParser.AddLabelContext):
+            """Dummy add label"""
+            expression = lambda self: (num(n1[0]), num(n2[0]))
+
+        class DummyBracketsLabel(blackbirdParser.BracketsLabelContext):
+            """Dummy brackets"""
+            def expression(self):
+                """Return expression"""
+                e = DummyAddLabel(parser, ctx)
+                e.PLUS = lambda: True
+                return e
+
+        class DummyPowerLabel(blackbirdParser.PowerLabelContext):
+            """Dummy power"""
+            expression = lambda self: (DummyBracketsLabel(parser, ctx), num(2))
+
+        expr = DummyPowerLabel(parser, ctx)
+        assert _expression(expr) == (n1[1]+n2[1])**2

```

Comment:
Seems like a very complicated series of steps were needed to just check if brackets are correctly parsed. Is there no more clear way to do it?